feat: self serve account deletion

joshsny opened the pull request:
## Problem

Users are not able to delete their account without contacting support. It takes up a long time for billing support.

Checks that they have no organization memberships or personal API keys before deleting. 

## Changes

Add a setting for users to delete their account in the app

![2025-03-19 17 18 55](https://github.com/user-attachments/assets/0bd43398-4ef4-4a5b-8867-8dd7836aeabe)

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Added unit tests for the delete endpoint for user, ran the frontend stuff locally

joshsny committed:
feat(wip): self serve account deletion

joshsny committed:
feat(wip): account deletion

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

joshsny committed:
feat(wip): self serve account deletion

joshsny committed:
Merge branch 'self-serve-account-deletion' of https://github.com/PostHog/posthog into self-serve-account-deletion

joshsny committed:
merge in master

joshsny committed:
Merge branch 'master' into self-serve-account-deletion

joshsny committed:
chore: allow deleting user accounts when user has no orgs

joshsny committed:
chore: update for non-owners

joshsny committed:
Merge branch 'master' into self-serve-account-deletion

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

joshsny committed:
feat(wip): deleting

joshsny committed:
chore: clean up logic

joshsny committed:
chore: update toasts

joshsny committed:
fix: type issues

joshsny committed:
chore: support redirect path

joshsny commented on frontend/src/scenes/settings/user/userDangerZoneLogic.ts:
```diff
@@ -0,0 +1,79 @@
+    listeners: () => ({
+        setDeleteUserModalOpen: ({ open }) => {
+            if (open) {
+                router.actions.replace(urls.settings('user-danger-zone'), { deletingUser: true })
```
We need to force this url in a few places, as we have to allow this page to be seen even when there is no active project

joshsny reviewed (commented):
This was a bit of a pain to get working due to performing actions outside of the current organization.

There are a couple existing issues I found here that will make the experience better:
1. We have an issue with error toasts showing due to API calls when there is no active project
2. There seems to be a permissions issue when deleting an org when the current user is an owner of that org but a member of the currently authenticated one

joshsny committed:
chore: add test for deleting another user

patricio-posthog reviewed (approved):
Approving the Python backend part -- just a minor suggestion but nothing that must be done before merging. Thanks for working on this!!

joshsny committed:
fix: add some more tests

zlwaterfield commented on posthog/api/user.py:
```diff
@@ -420,6 +420,18 @@ def get_serializer_context(self):
+
+        memberships = OrganizationMembership.objects.filter(user=user)
+
+        if memberships.count() > 0:
```
+1

joshsny replied:
will update this when I get back :)

zlwaterfield commented on frontend/src/scenes/settings/user/UserDangerZone.tsx:
```diff
@@ -0,0 +1,195 @@
+                {organizations.length > 0 && (
+                    <>
+                        <p className="text-danger font-semibold">
+                            You must leave or delete all organizations before deleting your account.
```
I think we also need to check for personal API keys and have them delete those as well.

joshsny replied:
I checked their model but it looks like the FK is set to cascade on delete?

joshsny committed:
merge in master

joshsny committed:
Merge branch 'master' into self-serve-account-deletion

joshsny committed:
chore: put no orgs requirement with a permission, add personal api keys to ui

joshsny committed:
fix: remove comment

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

joshsny committed:
fix: scene logic formatting

joshsny committed:
Merge branch 'self-serve-account-deletion' of https://github.com/PostHog/posthog into self-serve-account-deletion

joshsny commented:
So sorry that I haven't checked in on this one in such a long time - I've addressed the comments about moving the organization membership requirements into a permission - @zlwaterfield would you mind taking a look at that when you get a moment, as I don't have much experience with django permissions and this is a sensitive area.

zlwaterfield commented on posthog/api/user.py:
```diff
@@ -420,6 +421,15 @@ def get_serializer_context(self):
             "user_permissions": UserPermissions(cast(User, self.request.user)),
         }
 
+    def destroy(self, request, *args, **kwargs):
```
You shouldn't need to overwrite this method. Adding `mixins.DestroyModelMixin` to the view set will called the default destroy which calls `get_object` which calls `check_object_permissions`.

zlwaterfield replied:
Maybe just test to confirm but I'm fairly sure. 

Other than that this looks good.

joshsny replied:
I couldn't get it to work originally - but it looks like we were skipping `check_object_permissions` in our `get_object` so I've added it there to fix this and removed the `destroy` method

joshsny committed:
fix: add object permission checks to get_object on user api

joshsny committed:
Merge branch 'master' into self-serve-account-deletion

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

joshsny committed:
chore: add a couple more tests to cover bases

joshsny commented:
@zlwaterfield requesting another review here just in case there was something you wanted to look at before I merge

joshsny committed:
Merge branch 'self-serve-account-deletion' of https://github.com/PostHog/posthog into self-serve-account-deletion

joshsny committed:
Merge branch 'master' into self-serve-account-deletion

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

zlwaterfield commented on posthog/api/user.py:
```diff
@@ -399,6 +400,7 @@ def get_object(self) -> User:
         lookup_value = self.kwargs[self.lookup_field]
         request_user = cast(User, self.request.user)  # Must be authenticated to access this endpoint
         if lookup_value == "@me":
+            self.check_object_permissions(self.request, request_user)
```
This actually looks like it may have been a hole. Nice catch.

joshsny committed:
fix: revert unwanted changes to core filter definitions

joshsny commented on frontend/src/taxonomy/core-filter-definitions-by-group.json:
Not sure what happened there - I needed to `--no-verify` to get the commit through without it modifying this, must be something wrong with my linting setup on JSON files locally.

zlwaterfield reviewed (approved):
Tested locally, everything looks to be working well. Nice work. 

Impressive flow with multiple memberships and leaving/deleting.

benjackwhite commented on frontend/src/scenes/userLogic.ts:
```diff
@@ -76,6 +77,20 @@ export const userLogic = kea<userLogicType>([
                 },
+                deleteUser: async () => {
+                    return await api
+                        .delete('api/users/@me/')
```
This is potentially overkill but I have seen patterns before where its like "/request_deletion_code" which returns a value (like a time bound hashed sha of the user name or something) and then the actual request requires that code which can be good for essentially building double confirmation into the api itself.

This PR is so good that I don't think its necessary but food for thought in the future.

benjackwhite commented on frontend/src/scenes/userLogic.ts:
```diff
@@ -76,6 +77,20 @@ export const userLogic = kea<userLogicType>([
+                    return await api
+                        .delete('api/users/@me/')
+                        .then(() => {
+                            actions.deleteUserSuccess(null)
```
nit: Because this is a `loader` you dont need these manual actions. They will automatically get called if the async function resolves or throws. then you could move the actions.logout to the `deleteUserSuccess` listener which would be cleaner

benjackwhite reviewed (approved):
This is such a nice PR dude. 🙌

joshsny committed:
chore: use loader correctly

joshsny committed:
Merge branch 'master' into self-serve-account-deletion
